reader_writer_test.go

```diff
@@ -0,0 +1,482 @@
+	}
+
+	err = ionHashWriter.FieldName("b")
+	if err != nil {
+		t.Errorf("expected ionHashWriter.FieldName(\"b\") to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.WriteInt(1)
+	if err != nil {
+		t.Errorf("expected ionHashWriter.WriteInt(1) to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.EndStruct()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.EndStruct() to execute without errors; %s", err.Error())
+	}
+
+	err = writer.EndStruct()
+	if err != nil {
+		t.Errorf("expected writer.EndStruct() to execute without errors; %s", err.Error())
+	}
+
+	writeHash, err := ionHashWriter.Sum(nil)
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.Finish()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Finish() to execute without errors; %s", err.Error())
+	}
+
+	err = writer.Finish()
+	if err != nil {
+		t.Errorf("expected writer.Finish() to execute without errors; %s", err.Error())
+	}
+
+	reader := ion.NewReaderBytes(buf.Bytes())
+
+	hr, err := NewHashReader(reader, newIdentityHasherProvider())
+	if err != nil {
+		t.Fatalf("expected NewHashReader() to successfully create a HashReader; %s", err.Error())
+	}
+
+	ionHashReader, ok := hr.(*hashReader)
+	if !ok {
+		t.Fatalf("expected hr to be of type hashReader")
+	}
+
+	// We are reading the nested struct that we just wrote: {a:{b:1}}
+	// We use the ion reader to read the outer struct (ie. {a:_})
+	if !reader.Next() {
+		err = reader.Err()
+		if err != nil {
+			t.Errorf("expected reader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	err = reader.StepIn()
+	if err != nil {
+		t.Errorf("expected reader.StepIn() to execute without errors; %s", err.Error())
+	}
+
+	if !reader.Next() {
+		err = reader.Err()
+		if err != nil {
+			t.Errorf("expected reader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	// We use the ion hash reader to read the inner struct (ie. {b:1} inside {a:{b:1}})
+	err = ionHashReader.StepIn()
+	if err != nil {
+		t.Errorf("expected ionHashReader.StepIn() to execute without errors; %s", err.Error())
+	}
+
+	if !ionHashReader.Next() {
+		err = ionHashReader.Err()
+		if err != nil {
+			t.Errorf("expected ionHashReader.Next() to execute without errors; %s", err.Error())
+		}
+	}
+
+	err = ionHashReader.StepOut()
+	if err != nil {
+		t.Errorf("expected ionHashReader.StepOut() to execute without errors; %s", err.Error())
+	}
+
+	err = reader.StepOut()
+	if err != nil {
+		t.Errorf("expected reader.StepOut() to execute without errors; %s", err.Error())
+	}
+
+	sum, err := ionHashReader.Sum(nil)
+	if err != nil {
+		t.Fatalf("expected ionHashReader.Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	if !reflect.DeepEqual(sum, writeHash) {
+		t.Errorf("expected sum to be %v instead of %v", writeHash, sum)
+	}
+}
+
+func TestNoFieldNameInCurrentHash(t *testing.T) {
```

Comment:
The names of this test and `AssertNoFieldnameInCurrentHash` seem odd to me once I read their implementations. 
Maybe a better name could be `TestHashOfNestedValueInAStruct`. 

Is there a reason that we are testing for a value wrapped in a struct? Should we be testing values wrapped in other containers? I am asking in case I am missing something here, otherwise we could simplify the test a little.